trinity/sync/full/hexary_trie.py

```diff
@@ -0,0 +1,232 @@
+        """
+        pass
+
+    @property
+    def has_pending_requests(self) -> bool:
+        return len(self.requests) > 0
+
+    def next_batch(self, n: int = 1) -> List[SyncRequest]:
+        """Return the next requests that should be dispatched."""
+        if len(self.queue) == 0:
+            return []
+        batch = list(reversed((self.queue[-n:])))
+        self.queue = self.queue[:-n]
+        return batch
+
+    async def schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                       leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                       is_raw: bool = False) -> None:
+        """Schedule a request for the node with the given key."""
+        if node_key in self._existing_nodes:
+            self.logger.debug("Node %s already exists in db", encode_hex(node_key))
+            return
+        if await self.db.coro_exists(node_key):
+            self._existing_nodes.add(node_key)
+            self.logger.debug("Node %s already exists in db", encode_hex(node_key))
+            return
+        self._schedule(node_key, parent, depth, leaf_callback, is_raw)
+
+    def _schedule(self, node_key: Hash32, parent: SyncRequest, depth: int,
+                  leaf_callback: Callable[[bytes, 'SyncRequest'], Awaitable[None]],
+                  is_raw: bool = False) -> None:
+        if parent is not None:
+            parent.dependencies += 1
+
+        existing = self.requests.get(node_key)
+        if existing is not None:
+            self.logger.debug(
+                "Already requesting %s, will just update parents list", node_key)
+            existing.parents.append(parent)
+            return
+
+        request = SyncRequest(node_key, parent, depth, leaf_callback, is_raw)
+        # Requests get added to both self.queue and self.requests; the former is used to keep
+        # track which requests should be sent next, and the latter is used to avoid scheduling a
+        # request for a given node multiple times.
+        self.logger.debug("Scheduling retrieval of %s", encode_hex(request.node_key))
+        self.requests[request.node_key] = request
+        bisect.insort(self.queue, request)
+
+    async def process(self, results: List[Tuple[Hash32, bytes]]) -> None:
+        """Process request results.
+
+        :param results: A list of two-tuples containing the node's key and data.
+        """
+        for node_key, data in results:
+            request = self.requests.get(node_key)
+            if request is None:
+                # This may happen if we resend a request for a node after waiting too long,
+                # and then eventually get two responses with it.
+                self.logger.debug(
+                    "No SyncRequest found for %s, maybe we got more than one response for it",
+                    encode_hex(node_key))
+                return
+
+            if request.data is not None:
```

Comment:
I think I've got a solution that'll work.  Lets remove the `data` property from the `SyncRequest` object and just use a boolean property like `is_done`.  We can set `is_done` here in this method, and just pass the raw `data` value into `self.commit`.  It looks like there are no other places where `request.data` is accessed.  This has the added benefit of reducing the memory overhead of this class.

Reply:
We don't always commit at the end of `process()`, so we need to keep the data for a given request until we have committed all of its children; see my previous comment: https://github.com/ethereum/py-evm/pull/1124#discussion_r207978869

Reply:
Ah, missed that.  Ok, :shipit: unless you see a clean way around this.